src/Chunk.ts

```diff
@@ -441,20 +441,29 @@ export default class Chunk {
 					? '[name].js'
 					: '[name][extname].js'
 				: options.entryFileNames;
-			path = relative(
-				preserveModulesRelativeDir,
-				`${dirname(sanitizedId)}/${renderNamePattern(
-					pattern,
-					'output.entryFileNames',
-					{
-						ext: () => extension.substr(1),
-						extname: () => extension,
-						format: () => options.format as string,
-						name: () => this.getChunkName()
-					},
-					this.getChunkInfo.bind(this)
-				)}`
+			let currentDir = dirname(sanitizedId);
+			const fileName = renderNamePattern(
+				pattern,
+				'output.entryFileNames',
+				{
+					ext: () => extension.substr(1),
+					extname: () => extension,
+					format: () => options.format as string,
+					name: () => this.getChunkName()
+				},
+				this.getChunkInfo.bind(this)
 			);
+			if (options.preserveModulesRoot) {
+				const preserveModulesRoot = resolve(options.preserveModulesRoot);
+				if (currentDir.startsWith(preserveModulesRoot)) {
+					currentDir = resolve(
+						preserveModulesRelativeDir,
+						currentDir.slice(preserveModulesRoot.length).replace(/^\//, '')
+					);
+				}
+			}
+			const currentPath = `${currentDir}/${fileName}`;
+			path = relative(preserveModulesRelativeDir, currentPath);
```

Comment:
I am thinking, is this really all necessary? Wouldn't this be the same:

```js
if (currentDir.startsWith(preserveModulesRoot)) {
  path = currentDir.slice(preserveModulesRoot.length).replace(/^\//, '')
} else {
  path = relative(preserveModulesRelativeDir, `${currentDir}/${fileName}`);
}
```

or am I overlooking something here?

Reply:
In practice, this will probably will be the case, at least for right now. We would need to add a check somewhere to make sure `preserveModulesRelativeDir` is resolved as a parent directory to `preserveModulesRoot` for this to work out.